File: paper/simulations/core/target_rate.py
"""
Target-rate least-squares power allocation via KKT + Lambert W.

Solves:  min  sum_i (log2(1 + a_i P_i) - T_i)^2
         s.t. sum_i P_i <= P_tot,  P_i >= 0

Closed-form per-channel allocation (active set):
    P_i(lam) = max(0, (2 / (lam * C^2)) * W(lam * C^2 * 2^{T_i} / (2 * a_i)) - 1/a_i)
where C = ln(2), W = Lambert W principal branch.
"""

import numpy as np
from scipy.special import lambertw


C = np.log(2.0)  # ln(2)
C2 = C ** 2      # (ln 2)^2


def cap_power(a: np.ndarray, k: np.ndarray) -> np.ndarray:
    """Per-channel power cap: P_bar_i = (2^{T_i} - 1) / a_i."""
    return (2.0 ** k - 1.0) / a


def pi_lambda(lam: float, a: np.ndarray, k: np.ndarray) -> np.ndarray:
    """
    Compute per-channel power P_i(lambda) for a given dual variable lambda.

    For lambda > 0, uses the Lambert W closed form.
    For lambda = 0, returns the cap powers (all targets met exactly).
    """
    N = len(a)
    p_bar = cap_power(a, k)

    if lam <= 0.0:
        return p_bar.copy()

    alpha = lam * C2 / (2.0 * a)          # alpha_i = lam * C^2 / (2 a_i)
    arg = alpha * (2.0 ** k)              # argument to Lambert W
    w_val = np.real(lambertw(arg, k=0))   # principal branch W_0

    # y_i = W(arg) / alpha_i, then P_i = (y_i - 1) / a_i
    # Equivalent: P_i = (2 / (lam * C^2)) * W(arg) - 1/a_i
    P = (2.0 / (lam * C2)) * w_val - 1.0 / a

    # Project: clamp to [0, P_bar_i]
    P = np.clip(P, 0.0, p_bar)
    return P


def sum_power(lam: float, a: np.ndarray, k: np.ndarray) -> float:
    """Total power S(lambda) = sum_i P_i(lambda)."""
    return float(np.sum(pi_lambda(lam, a, k)))


def objective(P: np.ndarray, a: np.ndarray, k: np.ndarray) -> float:
    """Evaluate J = sum_i (log2(1 + a_i P_i) - T_i)^2."""
    r = np.log2(1.0 + a * P)
    return float(np.sum((r - k) ** 2))


def solve(a: np.ndarray, k: np.ndarray, P_tot: float,
          tol: float = 1e-10, max_iter: int = 200) -> dict:
    """
    Solve the target-rate power allocation problem via bisection on lambda.

    Parameters
    ----------
    a : array, shape (N,)
        Channel gain-to-noise ratios.
    k : array, shape (N,)
        Per-channel targets (bits/s/Hz).
    P_tot : float
        Total power budget.
    tol : float
        Bisection tolerance on |S(lambda) - P_tot|.
    max_iter : int
        Maximum bisection iterations.

    Returns
    -------
    dict with keys:
        'P' : optimal power allocation (N,)
        'lambda_star' : optimal dual variable
        'objective' : objective value J*
        'rates' : achieved rates r_i*
        'iterations' : number of bisection iterations
        'power_used' : total power used
    """
    a = np.asarray(a, dtype=float)
    k = np.asarray(k, dtype=float)
    N = len(a)

    p_bar = cap_power(a, k)
    sum_caps = float(np.sum(p_bar))

    # Case A: all targets achievable
    if P_tot >= sum_caps:
        P_star = p_bar.copy()
        return {
            'P': P_star,
            'lambda_star': 0.0,
            'objective': 0.0,
            'rates': k.copy(),
            'iterations': 0,
            'power_used': sum_caps,
        }

    # Case B: bisection on lambda
    # Find upper bound: need S(lam_hi) <= P_tot
    lam_hi = 1.0
    while sum_power(lam_hi, a, k) > P_tot:
        lam_hi *= 2.0
    lam_lo = 0.0

    iterations = 0
    for iterations in range(1, max_iter + 1):
        lam_mid = (lam_lo + lam_hi) / 2.0
        s = sum_power(lam_mid, a, k)
        if abs(s - P_tot) < tol:
            break
        if s > P_tot:
            lam_lo = lam_mid
        else:
            lam_hi = lam_mid

    P_star = pi_lambda(lam_mid, a, k)
    rates = np.log2(1.0 + a * P_star)

    return {
        'P': P_star,
        'lambda_star': lam_mid,
        'objective': objective(P_star, a, k),
        'rates': rates,
        'iterations': iterations,
        'power_used': float(np.sum(P_star)),
    }

File: src/utils/targetRate.ts
import { lambertW } from './lambertW';
import type { TargetRateResult, SweepPoint } from '../types';

const EPS = 1e-10;

/** Get ln(B) for the chosen log base */
function getC(logBase: 'ln' | 'log2'): number {
  return logBase === 'ln' ? 1 : Math.LN2;
}

/** Get B^k for the chosen log base */
function getBk(k: number, logBase: 'ln' | 'log2'): number {
  return logBase === 'ln' ? Math.exp(k) : Math.pow(2, k);
}

/**
 * Power cap per channel: P_i^cap = (B^k - 1) / a_i
 * This is the power needed so that r_i = k exactly.
 */
export function computeCap(ai: number, k: number, logBase: 'ln' | 'log2'): number {
  return (getBk(k, logBase) - 1) / ai;
}

/**
 * Compute optimal P_i for a given lambda (KKT multiplier).
 *
 * From KKT stationarity:
 *   P_i(lambda) = max(0, (2 / (lambda * C^2)) * W(lambda * C^2 * B^k / (2*a_i)) - 1/a_i)
 *
 * where C = ln(B), W is Lambert W principal branch.
 */
export function computePi(
  lambda: number,
  ai: number,
  k: number,
  logBase: 'ln' | 'log2',
): number {
  if (lambda <= 0) return computeCap(ai, k, logBase);

  const C = getC(logBase);
  const C2 = C * C;
  const Bk = getBk(k, logBase);
  const arg = lambda * C2 * Bk / (2 * ai);
  const w = lambertW(arg);
  const pi = (2 / (lambda * C2)) * w - 1 / ai;
  return Math.max(0, pi);
}

/**
 * Total power at a given lambda across all channels.
 */
function totalPowerAtLambda(
  lambda: number,
  channelGains: number[],
  k: number,
  logBase: 'ln' | 'log2',
): number {
  let total = 0;
  for (const ai of channelGains) {
    total += computePi(lambda, ai, k, logBase);
  }
  return total;
}

/**
 * Solve the target-rate power allocation problem:
 *   minimize sum_i (r_i - k)^2
 *   subject to: sum P_i <= P_tot, P_i >= 0
 *
 * Uses bisection on lambda (the KKT multiplier).
 */
export function solveTargetRate(
  channelGains: number[],
  totalPower: number,
  k: number,
  logBase: 'ln' | 'log2',
): TargetRateResult {
  const n = channelGains.length;
  const C = getC(logBase);
  const Bk = getBk(k, logBase);

  // Compute caps
  const caps = channelGains.map(ai => (Bk - 1) / ai);
  const sumCaps = caps.reduce((s, c) => s + c, 0);

  // Case A: enough power to hit target on all channels
  if (totalPower >= sumCaps - EPS) {
    const powers = caps.slice();
    const rates = new Array(n).fill(k);
    return {
      powers,
      rates,
      caps,
      lambda: 0,
      objective: 0,
      perChannelError: new Array(n).fill(0),
      powerUsed: sumCaps,
      leftover: totalPower - sumCaps,
      activeCount: n,
      caseA: true,
    };
  }

  // Case B: power-limited — bisection to find lambda*
  // At lambda=0, total power = sumCaps > Ptot
  // At lambda -> inf, total power -> 0
  // Find lambda such that totalPowerAtLambda(lambda) = Ptot

  let lo = 0;
  let hi = Math.max(...channelGains.map(ai => 2 * k * ai / C)) * 2; // safe upper bound

  // Ensure hi gives less power than Ptot
  while (totalPowerAtLambda(hi, channelGains, k, logBase) > totalPower) {
    hi *= 2;
  }

  // Bisection
  for (let iter = 0; iter < 200; iter++) {
    const mid = (lo + hi) / 2;
    const s = totalPowerAtLambda(mid, channelGains, k, logBase);
    if (Math.abs(s - totalPower) < EPS) {
      lo = mid;
      break;
    }
    if (s > totalPower) {
      lo = mid; // need more lambda to reduce power
    } else {
      hi = mid; // need less lambda to increase power
    }
  }

  const lambda = lo;
  const powers = channelGains.map(ai => computePi(lambda, ai, k, logBase));
  const rates = channelGains.map((ai, i) =>
    Math.log(1 + ai * powers[i]) / C
  );
  const perChannelError = rates.map(r => (r - k) * (r - k));
  const objective = perChannelError.reduce((s, e) => s + e, 0);
  const powerUsed = powers.reduce((s, p) => s + p, 0);
  const activeCount = powers.filter(p => p > EPS).length;

  return {
    powers,
    rates,
    caps,
    lambda,
    objective,
    perChannelError,
    powerUsed,
    leftover: totalPower - powerUsed,
    activeCount,
    caseA: false,
  };
}

/**
 * Generate sweep data: J vs Ptot for a range of total powers.
 */
export function generateSweepData(
  channelGains: number[],
  k: number,
  logBase: 'ln' | 'log2',
  ptotMax: number,
  numPoints: number = 200,
): SweepPoint[] {
  const n = channelGains.length;
  const Bk = getBk(k, logBase);
  const sumCaps = channelGains.reduce((s, ai) => s + (Bk - 1) / ai, 0);

  // Include exact threshold point
  const ptotValues = new Set<number>();
  for (let i = 0; i <= numPoints; i++) {
    ptotValues.add(Math.round((ptotMax * i) / numPoints * 1e8) / 1e8);
  }
  if (sumCaps <= ptotMax) {
    ptotValues.add(Math.round(sumCaps * 1e8) / 1e8);
  }

  const sorted = Array.from(ptotValues).sort((a, b) => a - b);
  const data: SweepPoint[] = [];

  for (const ptot of sorted) {
    const result = solveTargetRate(channelGains, ptot, k, logBase);
    const point: SweepPoint = {
      ptot,
      objective: result.objective,
      powerUsed: result.powerUsed,
    };
    for (let i = 0; i < n; i++) {
      point[`p${i}`] = result.powers[i];
      point[`r${i}`] = result.rates[i];
    }
    data.push(point);
  }

  return data;
}
